source/simulator/core.cc

```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+              newton_residual = std::sqrt(newton_residual_velo * newton_residual_velo + newton_residual_pres * newton_residual_pres);
+
+              // We may need to do a line search if the solution doesn't decrease the norm of the rhs. TODO: more documentation.
+              LinearAlgebra::BlockVector backup_linearization_point = current_linearization_point;
+
+              double test_newton_residual = 0;
+              double test_newton_residual_velo = 0;
+              double test_newton_residual_pres = 0;
+              double lambda = 1;
+              double alfa = 1e-4;
+              unsigned int line_search_iteration = 0;
+
+              do
+                {
+                  current_linearization_point = backup_linearization_point;
+
+                  LinearAlgebra::BlockVector search_direction = solution;
+
+                  search_direction *= lambda;
+
+
+                  current_linearization_point.block(introspection.block_indices.pressure) += search_direction.block(introspection.block_indices.pressure);
+                  current_linearization_point.block(introspection.block_indices.velocities) += search_direction.block(introspection.block_indices.velocities);
+
+                  assemble_newton_stokes_matrix = rebuild_stokes_preconditioner = false;
+                  rebuild_stokes_matrix = true;
+
+                  assemble_stokes_system();
+
+                  test_newton_residual_velo = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+                  test_newton_residual_pres = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+                  test_newton_residual = std::sqrt(test_newton_residual_velo * test_newton_residual_velo + test_newton_residual_pres * test_newton_residual_pres);
+
+                  if (test_newton_residual < (1.0 - alfa * lambda) * newton_residual || line_search_iteration >= parameters.max_newton_line_search_iterations || use_picard)
+                    {
+                      pcout << newton_residual/initial_newton_residual << "   Norm of rhs = " << newton_residual << ":" << test_newton_residual  << ", relative residual: " << newton_residual/initial_newton_residual << ", theta = " << parameters.newton_theta << std::endl;
+                      break;
+                    }
+                  else
+                    {
+
+                      pcout << "   Iteration " << line_search_iteration << ", with norm of the rhs " << test_newton_residual << " and going to "
+                            << (1.0 - alfa * lambda) * newton_residual << ", relative residual: " << newton_residual/initial_newton_residual << ", theta = " << parameters.newton_theta << std::endl;
+
+                      lambda = lambda * (2.0/3.0);// TODO: make a parameter out of this.
+                    }
+
+                  line_search_iteration++;
+                }
+              while (true);
```

Comment:
Without having looked at the actual code in question, I'm a huge fan of making things explicit: if you think that at some point in code, a particular condition should be true, then put that into an assertion. This way, you can let the code check that indeed everything is as you expect it to be, and it also serves as documentation for others reading the code later on to understand what you thought was the case here.